parser-typechecker/src/Unison/Codebase/Metadata.hs

```diff
@@ -43,7 +44,15 @@ insert :: (Ord a, Ord n) => (a, Type, Value) -> Star a n -> Star a n
 insert (a, ty, v) = Star3.insertD23 (a, ty, (ty,v))
 
 delete :: (Ord a, Ord n) => (a, Type, Value) -> Star a n -> Star a n
-delete (a, ty, v) = Star3.deleteD23 (a, ty, (ty,v))
+delete (a, ty, v) s = let
+  s' = Star3.deleteD3 (a, (ty,v)) s
+  -- if (ty,v) is the last metadata of type ty
+  -- we also delete (a, ty) from the d2 index
+  metadataByType = List.multimap (toList (R.lookupDom a (Star3.d3 s)))
+  in 
+  case Map.lookup ty metadataByType of
+    Just vs | all (== v) vs -> Star3.deleteD2 (a, ty) s' 
+    _ -> s'
```

Comment:
Found and fixed this bug as part of this PR. Previous implementation was just WRONG and would delete ALL the metadata of the same type as the argument passed to delete, like if you deleted one `Doc` you deleted all of them. This implementation fixes that. 

This fix is exercised by the transcript (I got it to fail in the transcript first, then made this fix).